### PR TITLE
Add crop function

### LINKED ISSUE #2554: Crop/edit image before adding to a note
Originally reported on Google Code with ID 1652

```
Let the user select a part of an image, instead of using the whole image.
```

Reported by `nicolas.raoul` on 2013-01-30 03:54:16

### PR DESCRIPTION
**Pull request by NightXlt**
Add crop function and adapt most of all smart phone. Fix an unused resources bug. one quit add note page without saving, but the resources would be still exist.This is unreasonable and wasteful for  sdcard space.

Will fix #2554

### CODE REVIEW

**Comment by mikehardy** [comment]
Hi there! Thanks for submitting this as a PR so we could see the proposed code.

I had some time this morning to review, there are some structural problems with the PR which I want to go over first before looking at the specific code - 

1.  The PR includes unrelated changes. Specifically, you have included others commits into this one - like my dependency updates. The correct git flow is to git rebase your branch while you are working, not to merge them in. Additionally, there are some whitespace changes that have nothing to do with functionality - those should not be in a PR unless you include them as a completely separate commit so whitespace changes and functionality changes aren't mixed. I need those other items cleared out of this PR before I can merge it. If you have them mixed in your commit history I'm not sure what the best method is other than to rebase on your branch, work through the conflicts, then force push to your branch on github so your PR has only your changes.

1.  We should not be creating cache directories - Android has cache directories as a system facility. Just get a handle from the system to the app's cache directory and use that. Further, when you create a temp file, it should automatically delete when the process exist. Finally, if you just use the same image file name each time (maybe deleting anything with that name first?), that would guarantee we never take up more than one image worth of space at a time, that the system should delete the file when we exit, and finally, since it is in the cache directory - if you mark it in the system correctly as "deletable when user needs space" - it will be auto-cleared. That should eliminate all problems with file orphaning

Once these two structural things are taken care of I'll look at the revised PR to see how the actual implementation goes. It sounds like you have tested it and it works, so just moving the working code around until it matches this style should be easy, esp since the cache file changes would eliminate a fair bit of code I think

**Comment by NightXlt** [comment]
Yeah， Thanks for replying. I am so sorry for ignoring this pull request comment.

Two replies:

1. I am sorry for the irrelevant commit code,I have modified that.
2. Trust me , I have tried my best to use previous cache directory. I use cache directory to be the  generated crop image, but select from gallery image crop works, take photo crop doesn't., then I look up numerous relevant code. I find a common feature—all of them use sdCard directory to be cache directory. I try that, and it works on different smartphone. If you have better solutions, I am glad to receive it.

**Comment by mikehardy** [comment]
Thanks @NightXlt - back in my action list - I will look at is soon as I can

Just based on your reply though, sounds like the cache stuff is a real irritating area! If there's nothing we can do, then that's the way it goes. I will look a bit though, to see.

**Comment by NightXlt** [comment]
Thank you so much. Hope a better solution. I know this one is not good, but this is only one could adapt most of all smartphone. I also hope we could solve this problem more gracefully.

**Comment by NightXlt** [comment]
Does it works?

**Comment by mikehardy** [comment]
I am so sorry @NightXlt - I only have time on the weekends and this Saturday was full of other things. I will check it tomorrow

**Comment by NightXlt** [comment]
There is no need to rush. We have plenty of time. I am just afraid you forget about this.Thank you for your hard work.  @mikehardy

File: AnkiDroid/src/main/res/values/16-multimedia-editor.xml [hunk 1]
```diff
@@ -108,4 +108,10 @@
 
 	<!-- Network failure -->
 	<string name="network_no_connection">No connection</string>
+
+    <!-- Crop function-->
+    <string name="crop_image" translatable="false">Do you want to crop this image?</string>
+    <string name="crop_button" translatable="false">Crop image</string>
+    <string name="save_dialog_content" translatable="false">Your current image size is %sMB that is more than default image limit 1MB. Do you want to crop this image?</string>
```
Comment (mikehardy) [inline AnkiDroid/src/main/res/values/16-multimedia-editor.xml]: Suggest: Current image is %sMB. Default image limit is 1MB. Do you want to crop it?

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/IFieldController.java [hunk 2]
```diff
@@ -63,10 +63,12 @@
     void onActivityResult(int requestCode, int resultCode, Intent data);
 
     // Called when the controller has stopped showing the field in favor of another one
-    void onFocusLost();
+    default void onFocusLost(){
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/IFieldController.java]: This is cool - that's a Java1.8 feature that did not use to work on all Android APIs, but I just verified that it did in the docs (I tested this on an Android 4.2 tablet even)

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java [hunk 3]
```diff
@@ -211,6 +213,12 @@ protected void done() {
                 if (!f.exists()) {
                     bChangeToText = true;
                 }
+
+                long length = f.length();
+                if (length > 1024 * 1024) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java]: I'd like for the limit to be a static constant at the top of the file if at all possible, defined just one place with a clear name and then just referenced

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java [hunk 4]
```diff
@@ -52,19 +55,29 @@
...
 
     private ImageView mImagePreview;
+    private LinearLayout mLinearLayout;
 
-    private String mTempCameraImagePath;
+    private String mTempImagePath;
+    private String mTempImagePrePath;//save the last path to prevent crop or take photo action canceled
     private DisplayMetrics mMetrics = null;
 
+    private int showCropButton = 0;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: The use of this makes me think it should be a boolean?

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java [hunk 5]
```diff
@@ -242,5 +258,167 @@ private void setPreviewImage(String imagePath, int maxsize) {
...
+
+
+    /**
+     * Get image uri that adapts various model
+     * @return image uri
+     */
+    private Uri getImageUri(Context context, Intent data) {
+        String imagePath = null;
+        Uri uri = data.getData();
+        if (Build.VERSION.SDK_INT >= 19) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: we are mixing version codes and version numbers here. I searched through the whole codebase and it looks like VERSION_CODES is majority. Could you use the VERSION_CODE?

**Comment by NightXlt** [comment]
Hi, I have finished my job, please review changes.

**Comment by mikehardy** [comment]
This looks good to me and tests out - it works for me on devices (an old API17 tablet even) plus emulators.

@timrae if you want to take a look at this? I think its important to note that most of the code that is ugly (in the sense it's not using app cache directories etc) was discussed in this PR, it apparently doesn't work for all the cases (select from gallery or take new photo) so you have to make your own directory and do it. 

**given that**, I don't have any objections here and I'd like to merge. 

What do you think?

**Comment by mikehardy** [comment]
Also, I should mention for benefit of @NightXlt that Tim has different things he looks for - I've been working on AnkiDroid about a year, and Tim has been working on it waaay longer so he knows so much more about the app and the users. I've tested the code and looked for things I know of, but please continue to be patient and work with us in case Tim has a completely different set of things. I really appreciate your effort! Thanks

**Comment by NightXlt** [comment]
yeah，I am aware that the code is ugly, so I hope we could find better solutions.  I would pay attention to this PR. I am truly  appreciate you for what you have done. @mikehardy

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java [hunk 6]
```diff
@@ -1064,6 +1067,7 @@ else if (fieldEditText.hasFocus()) {
                         fieldEditText.getText().append(field.getFormattedValue());
                     }
                     NoteService.saveMedia(col, (MultimediaEditableNote) mNote);
+                    addPath(field);
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java]: You're writing `addPath(field)` here, but on the line above it's already doing `NoteService.saveMedia(col, (MultimediaEditableNote) mNote);`, which is [saving the media to the media database](https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/java/com/ichi2/anki/servicelayer/NoteService.java#L124). Once the media is saved to the media database, you can't just go and simply delete the files from the filesystem, it needs to be deleted via the libanki api I guess.

What is the current behavior with respect to cleaning up media when the user aborts a note save? I.e. if the user adds a picture from the camera, goes back to the note editor, and then closes the note editor without saving, is the media deleted or does it still remain? I suspect it's not cleaned up, as that whole codebase is a total mess. If it's not being cleaned up already, then I think no need to clean it up in this PR either, better to make a separate issue and PR for that and do it properly and consistently.

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java [hunk 7]
```diff
@@ -52,19 +55,29 @@
...
+    private LinearLayout mLinearLayout;
 
-    private String mTempCameraImagePath;
+    private String mTempImagePath;
+    private String mTempImagePrePath;//save the last path to prevent crop or take photo action canceled
     private DisplayMetrics mMetrics = null;
 
+    private boolean showCropButton = false;
+
+    private static final String sAnkiCacheDirectory = Environment.getExternalStorageDirectory() + "/AnkiDroid/cache";
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: Let's just use the main internal cache? `getCacheDir()`

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java [hunk 8]
```diff
@@ -242,5 +270,167 @@ private void setPreviewImage(String imagePath, int maxsize) {
...
+
+    /**
+     * Get Uri based on current image path
+     * @return current image path's uri
+     */
+    private Uri getCropUri() {
+        File file = new File(mTempImagePath);
+        Uri uri;
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+            uri = FileProvider.getUriForFile(mActivity, mActivity.getApplicationContext().getPackageName() + ".apkgfileprovider", file);
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: What's going on here with `apkgfileprovider` etc?

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java [hunk 9]
```diff
@@ -242,5 +270,167 @@ private void setPreviewImage(String imagePath, int maxsize) {
...
+        }
+        return uri;
+    }
+
+
+    /**
+     * Get image uri that adapts various model
+     * @return image uri
+     */
+    private Uri getImageUri(Context context, Intent data) {
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: This seems unrelated to cropping functionality, can you please explain in detail?

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java (hunk 6 above)
Comment (NightXlt) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java]: When I get the cropped image, but I quit and not save this image. At this moment, We need delete resources.

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java (hunk 7 above)
Comment (NightXlt) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: I am so sure that getCacheDir doesn't work on various smartphone. At least , In Xiaomi and HuaWei. showCropButton field is for controlling the visibility for crop button. The crop button is only show for first.

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java (hunk 6 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java]: @NightXlt
Yes, I understand that. It doesn't answer my questions though.

@mikehardy do you have time to look into it?

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java (hunk 7 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: I've never heard of it not working, but if you saw a problem with it, you can alternatively use `getExternalCacheDir()`.

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java (hunk 6 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java]: Sure, I just checked. It looked to me like the basic image field controller has a flag for "has temporary media" since it can do an internal rotate/compress phase, but it only deletes that on import, if you bail on import it'll orphan the file I think. 

The initial copy from camera to temp file is clean though, it works similar to here but creates the file with File.createTempFile which has pretty strong guarantees in Android land on deletion when the process is killed https://github.com/ankidroid/Anki-Android/blob/090c130f326316a41aa1770357b239274f8639a1/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java#L113

Line 204 in that file should be altered to createTempFile also I believe? https://github.com/ankidroid/Anki-Android/blob/090c130f326316a41aa1770357b239274f8639a1/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java#L224

And if I understand these things correctly, all the file cleanup here could also be eliminated if any File creation here used the File.createTempFile API as well

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java (hunk 7 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: I'm surprised as well - and having investigated the temp file thing above it looks like they just use Activity.getCacheDir to do it, but maybe there's a FileProvider issue or something? I hate to contradict tested experience as it may just be that rotateAndCrop in existing code fails on those devices and we don't know it

https://github.com/ankidroid/Anki-Android/blob/090c130f326316a41aa1770357b239274f8639a1/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java#L112

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java (hunk 8 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: This is the same mechanism used in existing code from that line 112 reference above, it's line 116 in our codebase https://github.com/ankidroid/Anki-Android/blob/090c130f326316a41aa1770357b239274f8639a1/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java#L116

**Comment by mikehardy** [comment]
Okay you two - I added context to some of Tim's questions but they were not answers really - each one still needs to be addressed by @NightXlt but maybe @timrae my links to existing camera access -> take picture -> rotate and crop -> save workflow help?

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java (hunk 7 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: Let's use `getExternalCacheDir()`. If we use a cachedir I believe we shouldn't need to use a fileprovider.

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java (hunk 8 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: If we use a cachedir this shouldn't be necessary I believe.

File: AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java (hunk 6 above)
Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java]: AFAIK we're not talking about cleaning up temp media here, rather we're talking about cleaning up permanent media from the collection.media folder when the user closes the note editor without saving

Comment (timrae) [reply AnkiDroid/src/main/java/com/ichi2/anki/NoteEditor.java]: If we're going to clean up those media then I want to clean up all of them, not just some of them, and I want to clean them up correctly via libanki, not by manually deleting the file from filesystem.

**Comment by NightXlt** [comment]
I am traveling, I will reply in two weeks.

**Comment by NightXlt** [comment]
Actually, this waste resources is an extra existed problem, and I wanted to solve this problem moths ago. What I need to do further?

**Comment by mikehardy** [comment]
I'd love to see this added - it is a frequent request plus a pretty serious misuse of resources to push full-size camera photos out as media. It seems like communication is an issue. Maybe we could use the main PR textarea with a punchlist of minimum concrete actions to take in order to get it done? Having comments scattered around doesn't seem to be working, but the code itself was working when I tested it

**Comment by NightXlt** [comment]
I will explore several possibilities  of cache dir again.

**Comment by NightXlt** [comment]
All issue and answer of my code:
1. 
issue: Clean up permanent media 
answer: @timrae his/her viewpoint is right. I viewed some code about saving media in developing, but I found the media just moved from cache dir to collection.media dir, and I decide to delete media from unused  resources directly when I choose to quit and not save. If the media is saved into database, that  needs to be deleted from database first. But I am not familiar with your database operation.  
My personal solution is delete my relate code about cleaning up permanent media.

2.
issue: External cache directory
answer: @timrae  @mikehardy , I try to use application 's cache again. I get several familiar crashes that remind me why I choose external cache directory.  The fact as follows:
image
image
We can see that happens NPE. When using application's cache dir,  the system won't write cropped image into created file. However, If I use external dir ,that could. That is weird. I cannot say why does it happen. I guess that is android system bug.  I would be appreciate if you could find use cache solution that could implement crop function. 
My personal solution is keep the code.

**Comment by NightXlt** [comment]
@mikehardy  @timrae  what do you think of the above proposal?

**Comment by mikehardy** [comment]
Worth mentioning this is still a popular request, it just came up on the Play Store reviews again. I understand the location of file saving is a sticking point, but it appears from the responses that via direct experience the elegant ways of doing things show failure. So I'm not sure how to break the stalemate here.

**Comment by NightXlt** [comment]
could I use third-party library like uCrop？

**Comment by mikehardy** [comment]
uCrop looks fantastic actually - it would be great to have the majority of the code maintained by others to avoid duplication of effort, and that project looks like the best option. I took a look at the general state of "image cropping in android" and that's probably best basis to work on though all the libraries look poorly maintained. If any of them have a fork and become active though it'll be uCrop since it has the most activity is my guess.

Maybe that would be a way to break the logjam and move forward, I'd love to see an implementation using that - maybe a clean second PR that just integrates it?

**Comment by NightXlt** [comment]
@mikehardy  Sorry to tell you this, I have tried to use uCrop library, but that cannot work on the scenario, which cropping after taking photos would throw exception. 

```java
 E/TransformImageView: onFailure: setImageUri
    java.io.FileNotFoundException: open failed: ENOENT (No such file or directory)
```

I have committed issue in uCrop library, and they don't respond me until now.  It does not seems to be reliable. However I don't have any good idea to break through  the current cache directory predicament. I have tried many ideas to solve that crash when using caching dir. I am sorry to admit I am incapable to use cache dir to cropping img. I could crop img by using external dir. 

Sorry for not being able to help you.

**Comment by mikehardy** [comment]
A best effort is a best effort @NightXlt - perhaps @timrae can reconsider based on all the current information available. If not, I'll take a deeper dive into it. All your tests seem valid on inspection, and they were on real devices so I'm inclined to think this may just be a tough issue for the Android platform.

I'm also not surprised uCrop didn't get back to you - I am not sure why, but it appears that there are just no good / well-supported open cropping libraries out there. uCrop just seemed like the "least bad option", but didn't seem active when I checked it before. Maybe everyone just uses a system intent and expects the build in software to work and that's that.

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java  7
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/timrae/Anki-Android/pullRequest?prid=4361156&bid=14813999)

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java [hunk 10]
```diff
@@ -161,49 +183,58 @@ private DisplayMetrics getDisplayMetrics() {
...
 
 
-    private String rotateAndCompress(String inPath) {
+    private void rotateAndCompress() {
         // Set the rotation of the camera image and save as png
-        File f = new File(inPath);
+        File f = new File(mImagePath);
         // use same filename but with png extension for output file
-        String outPath = inPath.substring(0, inPath.lastIndexOf(".")) + ".png";
+        String outPath = mImagePath.substring(0, mImagePath.lastIndexOf(".")) + ".png";;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java]: Codacy Issue found: [An empty statement (semicolon) not part of a loop](https://app.codacy.com/app/timrae/Anki-Android/pullRequest?prid=4361156)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java  4
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4361156&bid=14813999)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java  4
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4361156&bid=14813999)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java  4
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4361156&bid=14813999)

**Comment by NightXlt** [comment]
## Purpose / Description
Add crop function 
## Fixes
External storage directory is used to store cropped image. Now I use external cache dir to be cache dir. Thx for your recommendation. @timrae . I don't notice the external cache dir in your comment. I finally know the difference between cache dir and external cache dir.

## Approach
Invoke system crop function to finish this feature. The core code is following:
```java
        Uri cropUri = Uri.fromFile(image);
        mField.setImagePath(mImagePath);
        mField.setHasTemporaryMedia(true);

        Intent intent = new Intent("com.android.camera.action.CROP");
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);
        }
        intent.setDataAndType(uri, "image/*");
        intent.putExtra("return-data", false);
        intent.putExtra(MediaStore.EXTRA_OUTPUT, cropUri);
        intent.putExtra("outputFormat", Bitmap.CompressFormat.PNG.toString());
        intent.putExtra("noFaceDetection", true); // no face detection
        mActivity.startActivityForResultWithoutAnimation(intent, ACTIVITY_CROP_PICTURE);
```

## How Has This Been Tested?
### Relevant phone
1. xiaomi 6, android 9, 
2. meizu 16s, android 9
### Test details
1. take photo/ select one image
2. Display image result, and show a dialog" Do you want to crop this image?"
3. Click sure to enter crop function
4. Crop image and click sure button
5. Display the cropped image normally，and show a CROP IMAGE button that used to crop image for users optionally. 
6. Click √ in the top right, which would check current image size whether it is bigger than 1MB.
If it is Bigger than 1 MB, it would be show a dialog to prompt user to cropp it.
7. Invoke `saveAndExit ()`method to save crop result in MultimediaEditFieldActivity file. It is a new method name by myself.
## Learning (optional, can help others)
1. `getImageUri（）` is a method for getting uri from Intent, and it adapts various phones' select image function's return intent. but I don't remember it clearly where I found that. I test it on ten phones and it actually works.
2. We need use different file name as  rotated image name every time. It is tested by myself.  First of all,  One image can be cropped many times. It seems to me that it is a write protection when I just decode a bitmap from path.  Prevent I change the file I just decode. I have to use different path instead of same path. So I create a method `CreateNewFile()`used to create a random file in external cache dir

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java  4
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4361156&bid=14813999)

**Comment by NightXlt** [comment]
@mikehardy @timrae Hello, guys. Could you please help review the code? I have reviewed my relevant code, and my PR request doc is above.

**Comment by mikehardy** [comment]
Hi @NightXlt - yes! I will review this in the next day or two (I'm a little busy with the holidays). This is still a very common feature request and I'd love to see it merge in, so I'm excited to look at the current version. Thanks!

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java  4
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4361156&bid=14813999)

**Comment by NightXlt** [comment]
> Hi @NightXlt - yes! I will review this in the next day or two (I'm a little busy with the holidays). This is still a very common feature request and I'd love to see it merge in, so I'm excited to look at the current version. Thanks!
eview
@mikehardy Excuse me, could you tell me does it work?

**Comment by mikehardy** [comment]
sorry @NightXlt I broke my review promise - I have it highlighted in my queue but have not had time to look at anything else deeply in AnkiDroid except a network change we have a deadline on, and a PR I've been trying to shape up for more than a year myself. I do want to get to this soon though, I don't have that many highlighted items so it will happen eventually. Thanks for your patience

**Comment by mikehardy** [comment]
@NightXlt you have been most patient! I just started handling releases for AnkiDroid and we have a pile of crash fixes I plan to release for 2.9, a quantity of new features that is already really large for 2.10, so I'm planning on getting those done shortly, and landing this on 2.11 as soon as 2.10 is cut. That means it will still be a little time but this is a very highly requested feature - it will go in eventually!

**Comment by NightXlt** [comment]
@mikehardy I'm glad to hear that. Patience is worth it.  If there is progress about this feature, sync it to me please. Thanks for your hark work, and highest respect to what you have done for anki.

**Comment by NightXlt** [comment]
@mikehardy Excuse me, I notice this PR hang around long times. Could I do something about this. Such as resolve conflicts.

**Comment by mikehardy** [comment]
Hey @NightXlt funny you mention that :-), we just finished a massive stability release series (2.9.x) and some internal / performance changes (2.10) and we are finally ready for some fresh features. The time has come to get this in! If you could rebase to current master and fix conflicts I would appreciate it

**Comment by NightXlt** [comment]
@mikehardy Sorry, I tried to resolve conflicts, but I failed. It's a hard nut, and I'm not available to resolve that. Maybe I will be available to retry this on SatSaturday.

**Comment by NightXlt** [comment]
@mikehardy  Done

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java [hunk 11]
```diff
@@ -22,23 +22,29 @@
 import android.Manifest;
 import android.content.Intent;
 import android.content.pm.PackageManager;
+import android.net.Uri;
 import android.os.Bundle;
 
 import androidx.annotation.NonNull;
 import androidx.annotation.VisibleForTesting;
 import androidx.core.app.ActivityCompat;
+import androidx.appcompat.widget.Toolbar;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java:31]: Codacy Issue found: [Avoid unused imports such as 'androidx.appcompat.widget.Toolbar'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4361156)

**Comment by timrae** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 2
           

Complexity increasing per file
==============================
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java  1
- AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicImageFieldController.java  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4361156&bid=14813999)

File: AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java [hunk 12]
```diff
@@ -48,7 +54,9 @@
 import com.ichi2.utils.Permissions;
 
 import java.io.File;
+import java.text.DecimalFormat;
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/activity/MultimediaEditFieldActivity.java:57]: Codacy Issue found: [Avoid unused imports such as 'java.text.DecimalFormat'](https://app.codacy.com/manual/timrae/Anki-Android/pullRequest?prid=4361156)

**Comment by mikehardy** [comment]
Thank you! It is my mission to get this in to 2.11 no matter what. I appreciate the effort

**Comment by mikehardy** [comment]
Carried this work into #6543 and built on it there, it's merged now. If you cannot claim the bounty for whatever reason on this @NightXlt let me know so I can if I can assist in anyway
